CDVD: Some Error handling, Status+Ready Flag changes and fix CdStop

refractionpcsx2 opened the pull request:
### Description of Changes
Adjusts the behaviour if the game request 0 or < 0 sectors

### Rationale behind Changes
This isn't right, but Armored Core 3 and Spyro Enter the Dragonfly kinda screw everything up, so this seems to resolve it.

### Suggested Testing Steps
boot games, if they work, we're good :)

Fixes #5155
Fixes issue with Spyro - Enter the Dragonfly TLB misses/failure to boot, patch still required however.
Fixes issue with bad sector LSN on Aberanbou Princess trial 
Animation Battle - Recca no Honou improved, still hangs in later FMV's without patch
Handling of bad sector counts in Evergrace demo and Armored Core 3

Savestate bump required, sorry

refractionpcsx2 committed:
CDVD: Zero sector handling and fix CdStop

gilderoylockhart01 commented:
Ok so this PR actually break more one already broken game. 

Abarenbou Princess (Taikenban) will produce a read timeout when loading the first level of the demo in the master branch, but the game will still continue regardless.

On this PR it will go back to BIOS before doing the read timeout, yeah refraction, fun.

Here is the blockdump: https://drive.google.com/file/d/1IGu4UR14_LzwJQ4m7sxXJKR7xdYNtAvk/view?usp=sharing

prafullpcsx2 commented:
@gilderoylockhart01  I have full version of the game and it doesn't suffer from this problem.

MrCK1 commented:
Tested a bunch of games, everything seems to work OK.

refractionpcsx2 committed:
CDVD: Add some error handling, fix up some IRQ stuff

refractionpcsx2 commented:
Added some small error handling and cleaned up some stuff which was incorrect, based on some new information, this should work nicely for games which try to read invalid sectors or a 0 or negative sector count (Armored Core 3 and the Evergrace demo)

refractionpcsx2 committed:
Core: Savestate version bump

refractionpcsx2 committed:
CDVD: Replace cdvd.Ready handling

refractionpcsx2 committed:
CDVD: Improve status handling/implement sticky status

gilderoylockhart01 commented:
Wait.... "Ecco The Dolphin no longer requires a patch".... what?

refractionpcsx2 commented:
> Wait.... "Ecco The Dolphin no longer requires a patch".... what?

One of the changes I did made it work, although I'm not entirely sure what I did is correct... I need to do some research.

gilderoylockhart01 commented:
Fixes Lego Indiana Jones non-fatal CDVD errors at boot.

refractionpcsx2 committed:
CDVD: Extra error handling and fix some flag details

refractionpcsx2 committed:
CDVD: Added missing busy/ready checks

refractionpcsx2 committed:
CDVD: Separate SCMD/NCMD Fifo's set to right size

refractionpcsx2 committed:
CDVD: Add more error checking, set seek status

refractionpcsx2 commented:
okay, that's a lot of changes, hopefully nothing is broken, please retest fixed/broken games.

Unfortunately no movement on Star Ocean (J) or Silent Hill 2

refractionpcsx2 committed:
CDVD: Add detection for PS2 Linux Disc 2

mirh commented:
So.. I feel like I'm on crazy pills because this is literally the bios I dumped myself and that AFAICT always worked just good 
Does it make any sense whatsoever that my `SCPH-77004_EUR_Con_0220_20060210_v15` just hangs to a black screen when booting to bios (meaning games also fail without fast boot), with **every** old pcsx2 build ***before*** this PR?
Another dump of a V12 had no such problem instead.

mirh commented:
Ok I found this was caused by the accompanying [.nvm file](https://github.com/user-attachments/files/30435735/SCPH-77004_EUR_Con_0220_20060210_v15_.F249B4A1.zip).
For some reason, unless I zero [bits](https://www.psdevwiki.com/ps2/index.php/MechaCon#Config_1_(OSD)) 0x2C1 (language?), 0x2C5 (time zone?) and 0x2CF of the NVRAM it hangs without this PR.
